common/src/main/scala/org/neo4j/spark/converter/DataConverter.scala

```diff
@@ -0,0 +1,181 @@
+import org.neo4j.spark.util.Neo4jUtil
+
+import java.time._
+import java.time.format.DateTimeFormatter
+import scala.annotation.tailrec
+import scala.collection.JavaConverters._
+
+trait DataConverter[T] {
+  def convert(value: Any, dataType: DataType = null): T
+
+  @tailrec
+  private[converter] final def extractStructType(dataType: DataType): StructType = dataType match {
+    case structType: StructType => structType
+    case mapType: MapType => extractStructType(mapType.valueType)
+    case arrayType: ArrayType => extractStructType(arrayType.elementType)
+    case _ => throw new UnsupportedOperationException(s"$dataType not supported")
+  }
+}
+
+object SparkToNeo4jDataConverter {
+  def apply(): SparkToNeo4jDataConverter = new SparkToNeo4jDataConverter()
+}
+
+class SparkToNeo4jDataConverter extends DataConverter[Value] {
+  override def convert(value: Any, dataType: DataType): Value = {
+    value match {
+      case date: java.sql.Date => convert(date.toLocalDate, dataType)
+      case timestamp: java.sql.Timestamp => convert(timestamp.toLocalDateTime, dataType)
+      case intValue: Int if dataType == DataTypes.DateType => convert(DateTimeUtils
+        .toJavaDate(intValue), dataType)
+      case longValue: Long if dataType == DataTypes.TimestampType => convert(DateTimeUtils
+        .toJavaTimestamp(longValue), dataType)
+      case unsafeRow: UnsafeRow => {
+        val structType = extractStructType(dataType)
+        val row = new GenericRowWithSchema(unsafeRow.toSeq(structType).toArray, structType)
+        convert(row)
+      }
+      case struct: GenericRowWithSchema => {
+        def toMap(struct: GenericRowWithSchema): Value = {
+          Values.value(
+            struct.schema.fields.map(
+              f => f.name -> convert(struct.getAs(f.name), f.dataType)
+            ).toMap.asJava)
+        }
+
+        try {
+          struct.getAs[UTF8String]("type").toString match {
+            case SchemaService.POINT_TYPE_2D => Values.point(struct.getAs[Number]("srid").intValue(),
+              struct.getAs[Number]("x").doubleValue(),
+              struct.getAs[Number]("y").doubleValue())
+            case SchemaService.POINT_TYPE_3D => Values.point(struct.getAs[Number]("srid").intValue(),
+              struct.getAs[Number]("x").doubleValue(),
+              struct.getAs[Number]("y").doubleValue(),
+              struct.getAs[Number]("z").doubleValue())
+            case SchemaService.DURATION_TYPE => Values.isoDuration(struct.getAs[Number]("months").longValue(),
+              struct.getAs[Number]("days").longValue(),
+              struct.getAs[Number]("seconds").longValue(),
+              struct.getAs[Number]("nanoseconds").intValue())
+            case SchemaService.TIME_TYPE_OFFSET => Values.value(OffsetTime.parse(struct.getAs[UTF8String]("value").toString))
+            case SchemaService.TIME_TYPE_LOCAL => Values.value(LocalTime.parse(struct.getAs[UTF8String]("value").toString))
+            case _ => toMap(struct)
```

Comment:
does it cover the case where `type` is not defined?

Reply:
yes